### DISCUSSION TITLE
Nanobind for Jit

### DISCUSSION THREAD

**Discussion by xelatihy** [wjakob/nanobind#52] [General]
I am exploring the idea of writing a very simple jit for teaching purposes.
The idea is to compile a python function to C++ code, compile that code
on the fly with a C++ compiler, and load the dll/so back interactively.
nanobind would serve to generate interface code.
This would work similarly to Nvidia's warp library.

I wonder whether there are examples of this use or reason for this not 
to work up front.

**Comment by wjakob** [comment]
Hi @xelatihy,

sorry about the late response, I missed this when you posted it initially. The general challenge is that you can't load the same extension module multiple times. You also can't unload and reload a binary extension module (which is something that _is_ possible for pure Python code). That doesn't need to be where it ends, and others have found workarounds.

For example, check out https://github.com/aldanor/ipybind

This is a jupyter notebook extension that on-the-fly compiles C++ extension modules via pybind11 that are imported into the Python process. The trick is just to give them some random name that changes every time.
